File: PyLibs/src/Libs/MsgRoute/MsgRouteHandleBase.py
# -*- coding: utf-8 -*-

from Libs.Core.GObject import GObject;
from Libs.DataStruct.MDictionary import MDictionary;
from Libs.EventHandle.AddOnceEventDispatch import AddOnceEventDispatch;

class MsgRouteHandleBase(GObject):
    def __init__(self):
        super(MsgRouteHandleBase, self).__init__();
        
        self.mTypeId = "MsgRouteHandleBase";
        
        self.mId2HandleDic = MDictionary();
        
        
    def addMsgRouteHandle(self, msgRouteID, handle):
        if(not self.mId2HandleDic.ContainsKey(msgRouteID)):
            self.mId2HandleDic[msgRouteID] = AddOnceEventDispatch();

        self.mId2HandleDic.value(msgRouteID).addEventHandle(None, handle);


    def removeMsgRouteHandle(self, msgRouteID, handle):
        if (self.mId2HandleDic.ContainsKey(msgRouteID)):
            self.mId2HandleDic.value(msgRouteID).removeEventHandle(None, handle);
          

    def handleMsg(self, msg):
        if (self.mId2HandleDic.ContainsKey(msg.m_msgID)):
            self.mId2HandleDic[msg.m_msgID].dispatchEvent(msg);
        else:
            # 输出日志
            pass;

File: JSLib/src/MsgRoute/MsgRouteHandleBase.js
using System.Collections.Generic;

namespace SDK.Lib
{
    public class MsgRouteHandleBase : GObject, ICalleeObject
    {
        public Dictionary<int, AddOnceEventDispatch> m_id2HandleDic;

        public MsgRouteHandleBase()
        {
            this.mTypeId = "MsgRouteHandleBase";

            m_id2HandleDic = new Dictionary<int, AddOnceEventDispatch>();
        }

        public void addMsgRouteHandle(MsgRouteID msgRouteID, MAction<IDispatchObject> handle)
        {
            if(!m_id2HandleDic.ContainsKey((int)msgRouteID))
            {
                m_id2HandleDic[(int)msgRouteID] = new AddOnceEventDispatch();
            }

            m_id2HandleDic[(int)msgRouteID].addEventHandle(null, handle);
        }

        public void removeMsgRouteHandle(MsgRouteID msgRouteID, MAction<IDispatchObject> handle)
        {
            if (m_id2HandleDic.ContainsKey((int)msgRouteID))
            {
                m_id2HandleDic[(int)msgRouteID].removeEventHandle(null, handle);
            }
        }

        public virtual void handleMsg(IDispatchObject dispObj)
        {
            MsgRouteBase msg = dispObj as MsgRouteBase;
            if (m_id2HandleDic.ContainsKey((int)msg.m_msgID))
            {
                m_id2HandleDic[(int)msg.m_msgID].dispatchEvent(msg);
            }
            else
            {
                Ctx.m_instance.m_logSys.log(string.Format(Ctx.m_instance.m_langMgr.getText(LangTypeId.eMsgRoute1, LangItemID.eItem1), (int)msg.m_msgID));
            }
        }

        public void call(IDispatchObject dispObj)
        {

        }
    }
}